File: apps/web/src/api/inventory.ts
import { fetchJson, fetchNoContent } from "./client";

export interface InventorySummary {
  totalQuantity: number;
  uniqueItems: number;
  uniqueParts: number;
}

export interface InventoryItem {
  partId: string;
  partName: string;
  category: string;
  colorCode: number;
  colorName: string;
  colorHex: string;
  alpha: number;
  quantity: number;
  renderAssetUrl: string | null;
  updatedAt: string;
  catalogAvailable: boolean;
}

export interface InventoryPage {
  items: InventoryItem[];
  page: number;
  pageSize: number;
  totalItems: number;
  totalPages: number;
}

export interface InventoryQuery {
  query: string;
  category: string;
  colorCode: number | null;
  page: number;
  pageSize?: number;
}

export function serializeInventoryQuery(input: InventoryQuery): string {
  const params = new URLSearchParams();
  if (input.query.trim()) params.set("query", input.query.trim());
  if (input.category) params.set("category", input.category);
  if (input.colorCode !== null) params.set("colorCode", String(input.colorCode));
  params.set("page", String(Math.max(1, Math.trunc(input.page))));
  if (input.pageSize !== undefined) params.set("pageSize", String(input.pageSize));
  return params.toString();
}

export function getInventorySummary(signal?: AbortSignal): Promise<InventorySummary> {
  return fetchJson<InventorySummary>("/api/inventory/summary", signal);
}

export function searchInventory(
  input: InventoryQuery,
  signal?: AbortSignal,
): Promise<InventoryPage> {
  return fetchJson<InventoryPage>(`/api/inventory/items?${serializeInventoryQuery(input)}`, signal);
}

export function getInventoryVariants(
  partId: string,
  signal?: AbortSignal,
): Promise<InventoryItem[]> {
  return fetchJson<InventoryItem[]>(`/api/inventory/items/${encodeURIComponent(partId)}`, signal);
}

export function setInventoryQuantity(
  partId: string,
  colorCode: number,
  quantity: number,
): Promise<InventoryItem> {
  return fetchJson<InventoryItem>(
    `/api/inventory/items/${encodeURIComponent(partId)}/${colorCode}`,
    undefined,
    {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ quantity }),
    },
  );
}

export function deleteInventoryItem(partId: string, colorCode: number): Promise<void> {
  return fetchNoContent(`/api/inventory/items/${encodeURIComponent(partId)}/${colorCode}`, {
    method: "DELETE",
  });
}

export type InventoryImportStrategy = "add" | "replace";
export type InventoryImportChange = "create" | "update" | "unchanged";
export type InventoryImportUnknownReason = "part" | "color" | "unmapped";
export type InventoryImportFormat = "native" | "rebrickable" | "bricklink" | "set";

export interface InventoryImportBucketSummary {
  rowCount: number;
  createCount: number;
  updateCount: number;
  unchangedCount: number;
  quantityDelta: number;
  missingPartCount: number;
  missingColorCount: number;
  missingMappingCount: number;
}

export interface InventoryImportPreviewRow {
  partId: string;
  sourcePartId: string;
  canonicalizedFrom: string | null;
  colorCode: number;
  quantity: number;
  currentQuantity: number;
  resultingQuantity: number;
  change: InventoryImportChange;
  unknownReason: InventoryImportUnknownReason | null;
  partName: string | null;
  colorName: string | null;
  colorHex: string | null;
  alpha: number | null;
  renderAssetUrl: string | null;
}

export interface InventoryImportRowIssue {
  lineNumber: number;
  code: string;
  message: string;
}

export interface InventoryImportPreview {
  fileName: string;
  format: InventoryImportFormat;
  strategy: InventoryImportStrategy;
  totalDataRows: number;
  plannedRowCount: number;
  duplicateRowCount: number;
  aliasCanonicalizedCount: number;
  ignoredColumns: string[];
  invalidRowCount: number;
  spareRowCount: number;
  mappingAvailable: boolean;
  known: InventoryImportBucketSummary;
  unknown: InventoryImportBucketSummary;
  rows: InventoryImportPreviewRow[];
  rowsTruncated: boolean;
  issues: InventoryImportRowIssue[];
  issuesTruncated: boolean;
  /** Only present for format === "set". */
  setNum: string | null;
  setName: string | null;
  officialPartCount: number | null;
  expandedQuantity: number | null;
}

export interface InventoryImportResult {
  format: InventoryImportFormat;
  strategy: InventoryImportStrategy;
  includeUnknown: boolean;
  appliedRowCount: number;
  createdCount: number;
  updatedCount: number;
  unchangedCount: number;
  skippedUnknownRowCount: number;
  invalidRowCount: number;
  quantityDelta: number;
  setNum: string | null;
  setName: string | null;
}

export function previewInventoryImport(
  file: File,
  strategy: InventoryImportStrategy,
  signal?: AbortSignal,
): Promise<InventoryImportPreview> {
  const body = new FormData();
  body.set("file", file);
  body.set("strategy", strategy);
  return fetchJson<InventoryImportPreview>("/api/inventory/import/preview", signal, {
    method: "POST",
    body,
  });
}

export function applyInventoryImport(
  file: File,
  strategy: InventoryImportStrategy,
  includeUnknown: boolean,
  format?: InventoryImportFormat,
): Promise<InventoryImportResult> {
  const body = new FormData();
  body.set("file", file);
  body.set("strategy", strategy);
  body.set("includeUnknown", includeUnknown ? "true" : "false");
  if (format !== undefined) body.set("format", format);
  return fetchJson<InventoryImportResult>("/api/inventory/import/apply", undefined, {
    method: "POST",
    body,
  });
}

export function previewSetImport(
  setNum: string,
  strategy: InventoryImportStrategy,
  signal?: AbortSignal,
): Promise<InventoryImportPreview> {
  return fetchJson<InventoryImportPreview>("/api/inventory/import/set/preview", signal, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ setNum, strategy }),
  });
}

export function applySetImport(
  setNum: string,
  strategy: InventoryImportStrategy,
  includeUnknown: boolean,
): Promise<InventoryImportResult> {
  return fetchJson<InventoryImportResult>("/api/inventory/import/set/apply", undefined, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ setNum, strategy, includeUnknown }),
  });
}

File: apps/api/app/schemas/inventory.py
from __future__ import annotations

from pydantic import Field

from app.schemas import CamelModel
from app.services.inventory_import import ChangeKind, Strategy, UnknownReason
from app.services.inventory_import_formats import ImportFormat


class ImportBucketSummaryResponse(CamelModel):
    row_count: int
    create_count: int
    update_count: int
    unchanged_count: int
    quantity_delta: int
    missing_part_count: int
    missing_color_count: int
    missing_mapping_count: int


class ImportPreviewRowResponse(CamelModel):
    part_id: str
    source_part_id: str
    canonicalized_from: str | None
    color_code: int
    quantity: int
    current_quantity: int
    resulting_quantity: int
    change: ChangeKind
    unknown_reason: UnknownReason | None
    part_name: str | None
    color_name: str | None
    color_hex: str | None
    alpha: int | None
    render_asset_url: str | None


class ImportRowIssueResponse(CamelModel):
    line_number: int
    code: str
    message: str


class InventoryImportPreviewResponse(CamelModel):
    file_name: str
    format: ImportFormat
    strategy: Strategy
    total_data_rows: int
    planned_row_count: int
    duplicate_row_count: int
    alias_canonicalized_count: int
    ignored_columns: list[str]
    invalid_row_count: int
    spare_row_count: int
    mapping_available: bool
    known: ImportBucketSummaryResponse
    unknown: ImportBucketSummaryResponse
    rows: list[ImportPreviewRowResponse]
    rows_truncated: bool
    issues: list[ImportRowIssueResponse]
    issues_truncated: bool
    # Only present for format == "set": the official set catalog vs. what
    # this app could actually expand (nested sub-inventories/minifigs are
    # not resolved, see docs/BULK_INVENTORY.md).
    set_num: str | None = None
    set_name: str | None = None
    official_part_count: int | None = None
    expanded_quantity: int | None = None


class InventoryImportApplyResponse(CamelModel):
    format: ImportFormat
    strategy: Strategy
    include_unknown: bool
    applied_row_count: int
    created_count: int
    updated_count: int
    unchanged_count: int
    skipped_unknown_row_count: int
    invalid_row_count: int
    quantity_delta: int
    set_num: str | None = None
    set_name: str | None = None


class SetImportPreviewRequest(CamelModel):
    set_num: str = Field(min_length=1, max_length=32)
    strategy: Strategy


class SetImportApplyRequest(CamelModel):
    set_num: str = Field(min_length=1, max_length=32)
    strategy: Strategy
    include_unknown: bool
